Enable `PLW1514` rule (use `utf-8` encoding)

Danipulok opened the pull request:
Hey!

**Summary**
I'm using Windows. While I was working on #3523, I had issues with rendering docs because of `docs/.hooks`

I decided to change some places inside `docs` and `clai` and `pydantic-evals`, so that they use `utf-8`.

Windows by default uses other encoding, and `utf-8` needs to be specified. 
(At least I really don't like this behavior at all, and it will be [changed](https://docs.python.org/3.15/whatsnew/3.15.html#whatsnew315-utf8-default) in python 3.15)

**Note**
There're some more places where `utf-8` can be specified.
I noticed in some tests you actually specify `utf-8`, so probably you want to minimize usage of `utf-8`
Then I can leave `utf-8` only in `.docs`, if it's acceptable

Viicos reviewed (commented):
Might be worth enabling the https://docs.astral.sh/ruff/rules/unspecified-encoding/ rule?

Danipulok commented:
@Viicos hey!  
Happy to enable it and lint the whole project if that works for you.  
Shall I do it?

Danipulok commented:
I updated the PR, please check
@Viicos, @DouweM 

**Changes:**
- Add `encoding='utf-8'` everywhere;
- Add `PLW1514` rule for `ruff`;
- Enable `preview` mode for `[tool.ruff]`;
- Refactored `docs/.hooks/test_snippets.py‎` to make them Windows-compatible;

**Notes:**
1. If `[tool.ruff.format]` `preview = false` is unset, about 75 files are changed via formatter;

2. `docs/.hooks/test_snippets.py‎` tests failed on Windows with:

```
            finally:
>               os.unlink(f.name)
E               PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\Dan\\AppData\\Local\\Temp\\tmpfbq5k65p.py'

docs\.hooks\test_snippets.py:104: PermissionError
                                                   Summary of Failures                                                    
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃  File                          ┃  Function                        ┃  Function Line  ┃  Error Line  ┃  Error            ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│  docs\.hooks\test_snippets.py  │  test_parse_file_sections_basic  │  81             │  104         │  PermissionError  │
└────────────────────────────────┴──────────────────────────────────┴─────────────────┴──────────────┴───────────────────┘
Results (0.67s):
         1 failed
        24 passed
```
Now they are working

3. I also removed `f.flush()` [link](https://github.com/pydantic/pydantic-ai/pull/3524/files#diff-0d79ffff9a61d24a24795adb45b7860915119e94f14cdacbdb9d8c156ad0dba1L90-L92) everywhere in `docs/.hooks/test_snippets.py‎`, because it had no effect when using `Path.write_text(...)` 

4. I also removed those [3 lines](https://github.com/pydantic/pydantic-ai/pull/3524/files#diff-0d79ffff9a61d24a24795adb45b7860915119e94f14cdacbdb9d8c156ad0dba1L537-L539) in `docs/.hooks/test_snippets.py‎`,, which did nothing

5. `PLW1514` does not work perfectly, so I had to manually search for `open`, `write_text` and `read_text`

Danipulok commented:
It also seems that pinned `ruff` version is old and does not catch most of the encoding errors (regarding my note 5)
Should I update it in this / other PR or no?

Viicos commented on pyproject.toml:
```diff
@@ -134,6 +134,7 @@ include = ["/README.md", "/Makefile", "/tests"]
 [tool.ruff]
 line-length = 120
 target-version = "py310"
+preview = true  # For `PLW1514`
```
Can be set in `[tool.ruff.lint]`, so that you don't need to override it in `format`.

Danipulok replied:
Updated, thank you!
Didn't think about it

Viicos commented:
> It also seems that pinned `ruff` version is old and does not catch most of the encoding errors (regarding my note 5)
> Should I update it in this / other PR or no?

I did the upgrade here: https://github.com/pydantic/pydantic-ai/pull/3547.

> `astral` team seems to be checking initial var's type, and if it's like this: `bundle_path: Path | None = None`, that it won't be considered as `Path`, even after `assert`s
> Also it seems they can't resolve some transformations like `Path(...) / "foo.py"`

Ruff currently isn't aware of the types of the variables, it only performs best effort single file analysis for now, so it is expected to get a bunch of false negatives.

Danipulok commented:
@Viicos, glad to help!
Thank you for merging
